### PR TITLE
istioctl: Use *mesh config* to find istiod XDS

### PR DESCRIPTION
**Pull request by esnible**
(Part of the work for https://github.com/istio/istio/issues/22274 )

Experimental implementation used for `istioctl experimental proxy-status` only.

If the user does supplies neither `--xds-port 15012` (self-hosted control plane) nor `--xds-address mycloud.com:port` (external control plane), _istioctl_ will use values in the injector config to default those flags.

Note: If the user is canarying they still need to set `--revision` at this time.

This should make things much easier for users with >1 mesh on >1 cluster.

This sets up some of the work needed before we merge the `x proxy-status` and non-experimental `proxy-status`.

### CODE REVIEW

File: tests/integration/pilot/istioctl_test.go [hunk 1]
```diff
@@ -373,6 +373,62 @@ func TestProxyStatus(t *testing.T) {
...
+				args = []string{
+					"experimenta", "proxy-status", fmt.Sprintf("%s.%s", podID, apps.Namespace.Name()),
+				}
+				output, _ = istioCtl.InvokeOrFail(t, args)
+				return expectSubstrings(output, "Clusters Match", "Listeners Match", "Routes Match")
+			})
+
+			// test the --file param
+			retry.UntilSuccessOrFail(t, func() error {
+				filename := "ps-configdump.json"
```
Comment (howardjohn) [inline tests/integration/pilot/istioctl_test.go:417]: use `filepath.Join(t.TempDir(), "ps-configdump.json")` please, or it ends up in the source directory. I sent a PR to fix the existing usage (which I assume you copied this from :slightly_smiling_face: ) already

**Review by shamsher31 — APPROVED** [review]

File: tests/integration/pilot/istioctl_test.go (hunk 1 above)
Comment (esnible) [reply tests/integration/pilot/istioctl_test.go]: Fixed

**Review by irisdingbj — APPROVED** [review]

File: tests/integration/pilot/istioctl_test.go (hunk 1 above)
Comment (esnible) [reply tests/integration/pilot/istioctl_test.go]: @howardjohn Please review again.

**Review by howardjohn — CHANGES_REQUESTED** [review]
I am a bit confused since the title says "injector" but code looks to be reading mesh config. This seems conflicting with plans in env wg: https://github.com/istio/istio/issues/31400.

The idea here is that external istio will read mesh config (including discovery address) from some local files. The in-cluster mesh config is just overrides, which presumably will not overwrite discoveryaddress (this may even be explicitly blocked).

**Comment by howardjohn** [comment]
cc @costinm ^

**Comment by esnible** [comment]
> I am a bit confused since the title says "injector" but code looks to be reading mesh config. This seems conflicting with plans in env wg: #31400.

@howardjohn @costinm OK.  Can anyone in @istio/wg-networking-maintainers or @istio/wg-environments-maintainers explain how to find

- The XDS address a pod is using
- The XDS address a namespace is using

I don't want to read the webhooks and spoof pod creation to figure out what address will be injected.

**Comment by howardjohn** [comment]
> The XDS address a pod is using

The only reliable way to know which address its using is to ask the pod. Which we currently don't have a way to do it but could add. You can attempt to reverse engineer it from various sources but it will never be 100% reliable.

> The XDS address a namespace is using

This is not answerable in general. We can get closer with "The XDS address a namespace is using _when a pod of revision X is injected via webhook_", but its still tricky. We could look at the webhook, but that isn't right either - the webhook could be in a different endpoint than XDS server (we are doing this today). Mesh config is also not safe since it may not live in the cluster.

Unfortunately I gave a lot of non answers... its a tricky problem. Maybe we need to add some additional info somewhere? For example, an `xdsAddress` annotation in the injector to handle the case of webhook url != xds address?

@costinm

**Comment by costinm** [comment]
Webhook is the worst option - we also support kube-inject ( user injected ). 

I would suggest:
- getting it from a pod - this will be the most reliable if you're trying to do something specific to a pod. Will work even if the pod is kube-injected. 
- mesh config - we should assume discovery address is there. But you may need to also read some of the env variables that control the root cert for the XDS server.
-

**Comment by costinm** [comment]
Also for operations specific to a pod - the user may use ProxyConfig annotations to override the discovery address. 
So I think getting it from a pod should be the first step. 

For operations like 'ps' - we may need to read all configmaps, so we can connect to each revision.

Another thing I want to raise is that we need to start limiting the scope of istioctl operations - right now it's done as 
mesh admin / all namespaces, but we need to restrict to either namespace or group of namespaces that the caller is authorized for ( using the JWT )

File: istioctl/pkg/xds/client.go [hunk 2]
```diff
@@ -93,3 +97,49 @@ func DialOptions(opts *clioptions.CentralControlPlaneOptions, kubeClient kube.Ex
...
+	// If the user did not specify any XDS address or Istiod selector, get them from injector config
+	if centralOpts.Xds == "" && centralOpts.XdsPodLabel == "" && centralOpts.XdsPodPort == 0 {
+		var meshConfig *v1alpha1.MeshConfig
+		var err error
+		if meshConfigFile != "" {
+			if meshConfig, err = mesh.ReadMeshConfig(meshConfigFile); err != nil {
+				return err
+			}
+		} else {
+			if meshConfig, err = clioptions.GetMeshConfigFromConfigMap(kubeClient, istioNamespace, revision); err != nil {
```
Comment (costinm) [inline istioctl/pkg/xds/client.go:112]: Certs will also need to be loaded at some point ( can be 'system' for external or meshes with real certs )

**Review by costinm — APPROVED** [review]
I think it's going into the wrong direction - despite the title of of the PR ( i don't see any injector )

**Comment by esnible** [comment]
@howardjohn @costinm I created [a document asking how to proceed](https://docs.google.com/document/d/18mFvj4j5pNOwWX2bDnRHt0XuwNvttxGcyzTH1GI_VZ4/edit#heading=h.xw1gqgyqs5b).  Please comment on it.  Which workgroup should I talk to?

Regarding the mesh config going away.  Perhaps it makes sense.  Users still want to know what their mesh settings are.  Last year I started a [_istioctl describe mesh_ document](https://docs.google.com/document/d/1_Y_tOZst7gtvqFbM6SnYHXsbbJMRD7XRPdcVsPedWag/edit#heading=h.px1wwffuiak4) but it went nowhere.  I'd like to bring it back.  Users may forget how they installed Istio, or it might have been installed for them by their cloud provider or IT department, and they might need to know the settings.  Please comment there.

**Pull request closed without merging**